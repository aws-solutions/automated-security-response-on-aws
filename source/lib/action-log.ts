// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Construct } from 'constructs';
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import { OrganizationPrincipal, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import * as cdk_nag from 'cdk-nag';
import { RemovalPolicy } from 'aws-cdk-lib';
import { OrgIdLookupConstruct } from './org-id-lookup';
import { addCfnGuardSuppression } from './cdk-helper/add-cfn-nag-suppression';

interface ActionLogProps {
  logGroupName: string;
}

export class ActionLog extends Construct {
  constructor(scope: Construct, id: string, props: ActionLogProps) {
    super(scope, id);
    const orgIdLookup = new OrgIdLookupConstruct(this, 'OrgIdLookup');

    // target LogGroup for ASR events. EventProcessor lambdas in member accounts will write filtered CloudTrail event here.
    const logGroup = new LogGroup(scope, 'CloudTrailEventsLogGroup', {
      logGroupName: props.logGroupName,
      retention: RetentionDays.ONE_YEAR,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    addCfnGuardSuppression(logGroup, 'CLOUDWATCH_LOG_GROUP_ENCRYPTED');

    // role for the EventProcessor lambda from the members to assume for cross-account write access to the LogGroup
    const crossAccountRole = new Role(this, 'CrossAccountLogWriterRole', {
      roleName: 'CrossAccountLogWriterRole',
      description: 'Role for cross-account access to write to CloudWatch Logs',
      assumedBy: new OrganizationPrincipal(orgIdLookup.organizationId),
    });
    crossAccountRole.addToPolicy(
      new PolicyStatement({
        actions: ['logs:CreateLogStream', 'logs:PutLogEvents', 'logs:DescribeLogStreams'],
        resources: [
          logGroup.logGroupArn,
          `${logGroup.logGroupArn}:*`, // Allow access to all log streams in this group
        ],
      }),
    );
    logGroup.grantWrite(crossAccountRole);

    cdk_nag.NagSuppressions.addResourceSuppressions(
      crossAccountRole,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Resource * is required to write to arbitrary LogStreams in this LogGroup',
        },
      ],
      true, // apply to children
    );
    addCfnGuardSuppression(crossAccountRole, 'CFN_NO_EXPLICIT_RESOURCE_NAMES');
  }
}
