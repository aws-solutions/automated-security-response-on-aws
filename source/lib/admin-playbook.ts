// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CfnCondition, CfnParameter, CfnResource, Fn, NestedStack, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface AdminPlaybookProps {
  name: string;
  stackDependencies?: CfnResource[];
  defaultState?: 'yes' | 'no';
  description?: string;
}

export class AdminPlaybook {
  parameterName = '';
  playbookStack: Stack;

  constructor(scope: Construct, props: AdminPlaybookProps) {
    const templateFile = `${props.name}Stack.template`;
    const illegalChars = /[\\._]/g;
    const playbookName = props.name.replace(illegalChars, '');
    this.parameterName = `Load${playbookName}AdminStack`;

    //---------------------------------------------------------------------
    // Playbook Template Nested Stack
    //
    const stackOption = new CfnParameter(scope, `LoadAdminStack${playbookName}`, {
      type: 'String',
      description:
        props.description ?? `Install the admin components for automated remediation of ${props.name} controls?`,
      default: props.defaultState ?? 'no',
      allowedValues: ['yes', 'no'],
    });
    stackOption.overrideLogicalId(this.parameterName);

    this.playbookStack = new NestedStack(scope, `PlaybookAdminStack${playbookName}`);
    const cfnStack = this.playbookStack.nestedStackResource as CfnResource;
    cfnStack.addPropertyOverride(
      'TemplateURL',
      'https://' +
        Fn.findInMap('SourceCode', 'General', 'S3Bucket') +
        '-reference.s3.amazonaws.com/' +
        Fn.findInMap('SourceCode', 'General', 'KeyPrefix') +
        '/playbooks/' +
        templateFile,
    );
    cfnStack.cfnOptions.condition = new CfnCondition(scope, `load${playbookName}Cond`, {
      expression: Fn.conditionEquals(stackOption, 'yes'),
    });
    props.stackDependencies?.forEach((dependency) => {
      cfnStack.node.addDependency(dependency);
    });
    cfnStack.overrideLogicalId(`PlaybookAdminStack${props.name}`);
  }
}
