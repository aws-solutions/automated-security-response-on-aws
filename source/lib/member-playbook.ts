// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CfnCondition, CfnParameter, CfnResource, Fn, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SerializedNestedStackFactory } from './cdk-helper/nested-stack';

export interface MemberPlaybookProps {
  readonly name: string;
  readonly nestedStackFactory: SerializedNestedStackFactory;
  readonly totalControls: number;
  readonly parameters?: Record<string, string>;
  readonly stackDependencies?: CfnResource[];
  readonly defaultState?: 'yes' | 'no';
  readonly description?: string;
  readonly stackLimit?: number;
}

export class MemberPlaybook {
  parameterName = '';
  playbookPrimaryStack: Stack;
  playbookOverflowStacks: Stack[];
  stackOption: CfnParameter;

  constructor(scope: Construct, props: MemberPlaybookProps) {
    const templateFile = `${props.name}MemberStack.template`;
    const logicalId = `PlaybookMemberStack${props.name}`;
    const illegalChars = /[\\._]/g;
    const playbookName = props.name.replace(illegalChars, '');
    this.parameterName = `Load${playbookName}MemberStack`;
    this.playbookOverflowStacks = [];

    //---------------------------------------------------------------------
    // Playbook Template Nested Stack
    //
    this.stackOption = new CfnParameter(scope, `LoadMemberStack${playbookName}`, {
      type: 'String',
      description:
        props.description ?? `Install the member components for automated remediation of ${props.name} controls?`,
      default: props.defaultState ?? 'no',
      allowedValues: ['yes', 'no'],
    });
    this.stackOption.overrideLogicalId(this.parameterName);

    this.playbookPrimaryStack = this.createPlaybookStack(templateFile, playbookName, logicalId, scope, props);

    if (props.stackLimit) {
      const numDivisions = Math.ceil(props.totalControls / props.stackLimit);
      for (let stackIndex = 1; stackIndex < numDivisions; stackIndex++) {
        const splitPlaybookStack = this.createPlaybookStack(
          `${props.name}MemberStack${stackIndex}.template`,
          `${playbookName}${stackIndex}`,
          `${logicalId}${stackIndex}`,
          scope,
          props,
        );
        this.playbookOverflowStacks.push(splitPlaybookStack);
      }
    }
  }

  private createPlaybookStack(
    templateFile: string,
    playbookName: string,
    logicalId: string,
    scope: Construct,
    props: MemberPlaybookProps,
  ) {
    const playbookStack = props.nestedStackFactory.addNestedStack(`PlaybookMemberStack${playbookName}`, {
      templateRelativePath: `playbooks/${templateFile}`,
      parameters: props.parameters,
      condition: new CfnCondition(scope, `load${playbookName}Cond`, {
        expression: Fn.conditionEquals(this.stackOption, 'yes'),
      }),
    });

    const cfnStack = playbookStack.nestedStackResource as CfnResource;
    cfnStack.overrideLogicalId(logicalId);
    return playbookStack;
  }
}
