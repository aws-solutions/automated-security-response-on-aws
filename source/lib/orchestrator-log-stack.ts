// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as cdk from 'aws-cdk-lib';
import { LogGroup, CfnLogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Key } from 'aws-cdk-lib/aws-kms';

export interface OrchLogStackProps extends cdk.StackProps {
  logGroupName: string;
  solutionId: string;
}

export class OrchLogStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: OrchLogStackProps) {
    super(scope, id, props);
    const stack = cdk.Stack.of(this);

    const reuseOrchLogGroup = new cdk.CfnParameter(this, 'Reuse Log Group', {
      type: 'String',
      description: `Reuse existing Orchestrator Log Group? Choose "yes" if you have deployed this stack before, otherwise "no"`,
      default: 'no',
      allowedValues: ['yes', 'no'],
    });
    reuseOrchLogGroup.overrideLogicalId(`ReuseOrchestratorLogGroup`);

    const kmsKeyArn = new cdk.CfnParameter(this, 'KMS Key Arn', {
      type: 'String',
      description: `ARN of the KMS key to use to encrypt log data.`,
    });
    kmsKeyArn.overrideLogicalId(`KmsKeyArn`);

    /**********************
     * Encrypted log group
     */
    // As of March 2021, CWLogs encryption is not yet supported in GovCloud
    // Choose based on partition

    const kmsKey = Key.fromKeyArn(this, 'KmsKey', kmsKeyArn.valueAsString);
    const orchestratorLogGroupEncrypted: LogGroup = new LogGroup(this, 'Orchestrator-Logs-Encrypted', {
      logGroupName: props.logGroupName,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
      retention: RetentionDays.TEN_YEARS,
      encryptionKey: kmsKey,
    });

    /************************
     * Unencrypted log group
     */
    const orchestratorLogGroupNOTEncrypted: LogGroup = new LogGroup(this, 'Orchestrator-Logs', {
      logGroupName: props.logGroupName,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
      retention: RetentionDays.TEN_YEARS,
    });

    /*******************
     *  Conditions
     */
    const isNotGovCloud = new cdk.CfnCondition(this, 'isNotGovCloud', {
      expression: cdk.Fn.conditionNot(cdk.Fn.conditionEquals(stack.partition, 'aws-us-gov')),
    });
    {
      const childToMod = orchestratorLogGroupEncrypted.node.defaultChild as CfnLogGroup;
      childToMod.cfnOptions.condition = new cdk.CfnCondition(this, 'Encrypted Log Group', {
        expression: cdk.Fn.conditionAnd(isNotGovCloud, cdk.Fn.conditionEquals(reuseOrchLogGroup.valueAsString, 'no')),
      });
    }

    {
      const childToMod = orchestratorLogGroupNOTEncrypted.node.defaultChild as CfnLogGroup;
      childToMod.cfnOptions.condition = new cdk.CfnCondition(this, 'Unencrypted Log Group', {
        expression: cdk.Fn.conditionAnd(
          cdk.Fn.conditionNot(isNotGovCloud),
          cdk.Fn.conditionEquals(reuseOrchLogGroup.valueAsString, 'no'),
        ),
      });
      childToMod.cfnOptions.metadata = {
        cfn_nag: {
          rules_to_suppress: [
            {
              id: 'W84',
              reason: 'KmsKeyId is not supported in GovCloud.',
            },
          ],
        },
      };
    }
  }
}
