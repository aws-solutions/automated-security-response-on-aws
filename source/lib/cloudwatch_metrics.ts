// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CfnCondition, CfnParameter, Duration, Fn } from 'aws-cdk-lib';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';
import setCondition from './cdk-helper/set-condition';
import {
  Color,
  ComparisonOperator,
  Dashboard,
  GraphWidget,
  GraphWidgetView,
  IMetric,
  LogQueryWidget,
  MathExpression,
  Metric,
  SingleValueWidget,
  TextWidget,
  TreatMissingData,
} from 'aws-cdk-lib/aws-cloudwatch';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { Key } from 'aws-cdk-lib/aws-kms';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { SC_REMEDIATIONS } from '../playbooks/SC/lib/sc_remediations';
import { IControl } from './sharrplaybook-construct';
import { addCfnGuardSuppression } from './cdk-helper/add-cfn-nag-suppression';

export interface CloudWatchMetricsProps {
  solutionId: string;
  schedulingQueueName: string;
  orchStateMachineArn: string;
  kmsKey: Key;
  actionLogLogGroupName: string;
  enhancedMetricsEnabled: CfnCondition;
}

export class CloudWatchMetrics {
  private readonly parameters: CfnParameter[] = [];
  private readonly useCloudWatchMetrics: CfnParameter;
  private readonly isUsingCloudWatchMetrics: CfnCondition;
  private readonly isUsingCloudWatchMetricsAlarms: CfnCondition;
  private readonly enhancedMetricsEnabled: CfnCondition;
  private readonly enhancedAlarmsEnabled: CfnCondition;

  constructor(scope: Construct, props: CloudWatchMetricsProps) {
    const RESOURCE_PREFIX = props.solutionId.replace(/^DEV-/, ''); // prefix on every resource name

    props.kmsKey.grantEncryptDecrypt(new ServicePrincipal('cloudwatch.amazonaws.com'));

    /// CloudWatch Metrics
    this.useCloudWatchMetrics = new CfnParameter(scope, 'UseCloudWatchMetrics', {
      type: 'String',
      description:
        'Enable collection of operational metrics and create a CloudWatch dashboard to monitor solution operations',
      default: 'yes',
      allowedValues: ['yes', 'no'],
    });
    this.parameters.push(this.useCloudWatchMetrics);

    this.isUsingCloudWatchMetrics = new CfnCondition(scope, 'isUsingCloudWatchMetrics', {
      expression: Fn.conditionEquals(this.useCloudWatchMetrics, 'yes'),
    });

    const useCloudWatchMetricsAlarms = new CfnParameter(scope, 'UseCloudWatchMetricsAlarms', {
      type: 'String',
      description: 'Create CloudWatch Alarms for gathered metrics',
      default: 'yes',
      allowedValues: ['yes', 'no'],
    });
    this.parameters.push(useCloudWatchMetricsAlarms);

    this.isUsingCloudWatchMetricsAlarms = new CfnCondition(scope, 'isUsingCloudWatchMetricsAlarms', {
      expression: Fn.conditionAnd(this.isUsingCloudWatchMetrics, Fn.conditionEquals(useCloudWatchMetricsAlarms, 'yes')),
    });

    this.enhancedMetricsEnabled = props.enhancedMetricsEnabled;
    this.enhancedAlarmsEnabled = new CfnCondition(scope, 'enhancedAlarmsEnabled', {
      expression: Fn.conditionAnd(this.enhancedMetricsEnabled, this.isUsingCloudWatchMetricsAlarms),
    });

    const remediationFailureAlarmThreshold = new CfnParameter(scope, 'RemediationFailureAlarmThreshold', {
      type: 'Number',
      description:
        'Percentage of failures in one period (1 day) to trigger the remediation failures alarm for a given control ID. E.g., to specify 20% then enter the number 20. These alarms will not be created if you select "no" on either of the following parameters: UseCloudWatchMetricsAlarms, EnableEnhancedCloudWatchMetrics.',
      default: 5,
    });
    this.parameters.push(remediationFailureAlarmThreshold);

    const sendCloudwatchMetricsParameter = new StringParameter(scope, 'ASR_SendCloudWatchMetrics', {
      description: 'Flag to enable or disable sending cloudwatch metrics.',
      parameterName: '/Solutions/' + RESOURCE_PREFIX + '/sendCloudwatchMetrics',
      stringValue: 'yes',
    });
    setCondition(sendCloudwatchMetricsParameter, this.isUsingCloudWatchMetrics);

    const defaultDuration = Duration.days(1);

    const lambdaErrorMetric = new Metric({
      namespace: 'ASR',
      metricName: 'RemediationOutcome',
      statistic: 'Sum',
      period: defaultDuration,
      dimensionsMap: { Outcome: 'LAMBDAERROR' },
      label: 'Lambda Error',
    });

    const remediationNotActiveErrorMetric = new Metric({
      namespace: 'ASR',
      metricName: 'RemediationOutcome',
      statistic: 'Sum',
      period: defaultDuration,
      dimensionsMap: { Outcome: 'REMEDIATIONNOTTACTIVE' },
      label: 'Remediation Not Active',
    });

    const noRemediationErrorMetric = new Metric({
      namespace: 'ASR',
      metricName: 'RemediationOutcome',
      statistic: 'Sum',
      period: defaultDuration,
      dimensionsMap: { Outcome: 'NOREMEDIATION' },
      label: 'No Remediation',
    });

    const standardNotEnabledErrorMetric = new Metric({
      namespace: 'ASR',
      metricName: 'RemediationOutcome',
      statistic: 'Sum',
      period: defaultDuration,
      dimensionsMap: { Outcome: 'STANDARDNOTENABLED' },
      label: 'Standard Not Enabled',
    });

    const automationDocumentFailedMetric = new Metric({
      namespace: 'ASR',
      metricName: 'RemediationOutcome',
      statistic: 'Sum',
      period: defaultDuration,
      dimensionsMap: { Outcome: 'FAILED' },
      label: 'SSM Doc Failed',
    });

    const successMetric = new Metric({
      namespace: 'ASR',
      metricName: 'RemediationOutcome',
      statistic: 'Sum',
      period: Duration.days(90), // 3 months
      dimensionsMap: { Outcome: 'SUCCESS' },
      label: 'Successful Remediations',
    });

    const hoursSavedMetric = new MathExpression({
      label: 'Estimated Hours Saved',
      period: defaultDuration,
      expression: '(m1 * 30) / 60',
      usingMetrics: {
        ['m1']: successMetric,
      },
    });

    const failuresByTypeExpression = new MathExpression({
      label: 'FAILURE',
      period: defaultDuration,
      expression: 'SUM([m1+m2+m3+m4+m5])',
      usingMetrics: {
        ['m1']: lambdaErrorMetric,
        ['m2']: remediationNotActiveErrorMetric,
        ['m3']: noRemediationErrorMetric,
        ['m4']: standardNotEnabledErrorMetric,
        ['m5']: automationDocumentFailedMetric,
      },
    });

    const remediationFailureRateExpression = new MathExpression({
      label: 'Overall Failure Rate',
      period: defaultDuration,
      expression: '(failuresByType / (failuresByType + successMetric)) * 100',
      usingMetrics: {
        ['failuresByType']: failuresByTypeExpression,
        ['successMetric']: successMetric,
      },
    });

    const failedAssumeRoleMetric = new Metric({
      namespace: 'ASR',
      metricName: 'AssumeRoleFailure',
      statistic: 'Sum',
      period: defaultDuration,
      label: 'Runbook Assume Role Failures',
    });

    /// CloudWatch Alarms
    const snsAlarmTopic = new Topic(scope, 'ASR-Alarm-Topic', {
      displayName: 'ASR Alarm Topic (' + RESOURCE_PREFIX + ')',
      topicName: RESOURCE_PREFIX + '-ASR_Alarm_Topic',
      masterKey: props.kmsKey,
    });
    setCondition(snsAlarmTopic, this.isUsingCloudWatchMetricsAlarms);

    const noRemediationErrorAlarm = noRemediationErrorMetric.createAlarm(scope, 'NoRemediationErrorAlarm', {
      alarmName: 'ASR-NoRemediation',
      evaluationPeriods: 1,
      threshold: 1,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      alarmDescription:
        'Remediation failed with NOREMEDIATION result. This indicates a remediation was attempted for an unsupported remediation',
      treatMissingData: TreatMissingData.NOT_BREACHING,
      datapointsToAlarm: 1,
      actionsEnabled: true,
    });
    setCondition(noRemediationErrorAlarm, this.isUsingCloudWatchMetricsAlarms);
    noRemediationErrorAlarm.addAlarmAction(new SnsAction(snsAlarmTopic));
    addCfnGuardSuppression(noRemediationErrorAlarm, 'CFN_NO_EXPLICIT_RESOURCE_NAMES');

    const failedAssumeRoleAlarm = failedAssumeRoleMetric.createAlarm(scope, 'FailedAssumeRoleAlarm', {
      alarmName: 'ASR-RunbookAssumeRoleFailure',
      evaluationPeriods: 1,
      threshold: 1,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      alarmDescription:
        'ASR Runbook Failed to assume role in an account. This indicates that a remediation was attempted in an account that does not have ASR deployed.',
      treatMissingData: TreatMissingData.NOT_BREACHING,
      datapointsToAlarm: 1,
      actionsEnabled: true,
    });
    setCondition(failedAssumeRoleAlarm, this.isUsingCloudWatchMetricsAlarms);
    failedAssumeRoleAlarm.addAlarmAction(new SnsAction(snsAlarmTopic));
    addCfnGuardSuppression(failedAssumeRoleAlarm, 'CFN_NO_EXPLICIT_RESOURCE_NAMES');

    const controlIds = SC_REMEDIATIONS.map((remediation: IControl) => remediation.control);
    const failureRateMetricsByControlId: IMetric[] = this.createAlarmsByControlId(
      scope,
      remediationFailureAlarmThreshold.valueAsNumber,
      snsAlarmTopic,
      defaultDuration,
      controlIds,
    );

    /// CloudWatch Dashboard
    const remediationDashboard = new Dashboard(scope, 'RemediationDashboard', {
      dashboardName: 'ASR-Remediation-Metrics-Dashboard',
      defaultInterval: Duration.days(7),
    });
    setCondition(remediationDashboard, this.isUsingCloudWatchMetrics);

    remediationDashboard.addWidgets(
      new TextWidget({
        markdown: `
## Total Successful Remediations
This widget displays the total number of successful remediations executed and total developer hours saved in the last 3 months.

We estimate that, on average, it takes 30 minutes of developer time to investigate & remediate a Security Hub finding. The "Estimated Hours Saved" widget uses this to estimate how many developer hours were saved by using ASR in the last 3 months.
`,
        height: 3,
        width: 24,
      }),
      new SingleValueWidget({
        title: 'Total Successful Remediations',
        metrics: [successMetric],
        setPeriodToTimeRange: true,
        height: 6,
      }),
      new SingleValueWidget({
        title: 'Estimated Hours Saved',
        metrics: [hoursSavedMetric],
        setPeriodToTimeRange: true,
        height: 6,
      }),
    );

    remediationDashboard.addWidgets(
      new TextWidget({
        markdown: `
## Remediation Failures by Type
This widget displays the frequency of various remediation failures. 
* \`Lambda Error\`: One or more of the solution's Lambda Functions failed to execute. See the Orchestrator step function execution for details.
* \`Remediation Not Active\`: The runbook associated with this remediation is not properly deployed in the solution's Admin and/or Member stack. Verify the solution's parameters.
* \`No Remediation\`: ASR does not currently implement a remediation for the executed finding.
* \`Standard Not Enabled\`: The Security Standard associated with the finding is not enabled in Security Hub. Navigate to the Security Hub console to activate the Standard.
* \`SSM Doc Failed\`: The remediation script failed to execute. Check the Orchestrator step function to determine which account the remediation was executed in, then view the SSM automation execution history for failures.

If there is an increase in \`NOREMEDIATION\` results, this indicates that remediations are being attempted for remediations not currently included in ASR. You should verify that this is not caused by a modified automatic remediation EventBridge rule.
`,
        height: 6,
        width: 24,
      }),
      new GraphWidget({
        title: 'Remediation Failures',
        left: [failuresByTypeExpression],
        leftYAxis: {
          showUnits: false,
        },
      }),
      new GraphWidget({
        title: 'Remediation Failures by Type',
        left: [
          lambdaErrorMetric,
          remediationNotActiveErrorMetric,
          noRemediationErrorMetric,
          standardNotEnabledErrorMetric,
          automationDocumentFailedMetric,
        ],
        leftYAxis: {
          showUnits: false,
        },
      }),
      new GraphWidget({
        title: 'Remediation Failure Rate',
        left: [remediationFailureRateExpression],
        leftYAxis: {
          showUnits: false,
        },
      }),
    );
    remediationDashboard.addWidgets(
      new TextWidget({
        markdown: `
## Remediation Success/Failure by Control ID
This widget displays the number of successful and failed remediations by Control ID. You must select "Yes" for EnableEnhancedCloudWatchMetrics when deploying the Admin stack to view these metrics.

The number of failed remediations per Control ID can inform you of frequent issues that arise when ASR attempts to remediate a specific finding in your AWS environment. If a high number of failures occur on a small subset of controls, you can investigate the issue by navigating to \`Systems Manager > Automation\` to search for recent executions of the control runbook associated with the failing Control ID. 
`,
        height: 3,
        width: 24,
      }),
      new GraphWidget({
        title: 'Successful remediations by Control Id',
        left: [
          new MathExpression({
            expression: "SORT(SEARCH('{ASR,ControlId,Outcome} Outcome=\"SUCCESS\"', 'Sum'), SUM, ASC)",
            label: '',
            period: defaultDuration,
          }),
        ],
        view: GraphWidgetView.BAR,
        statistic: 'Sum',
      }),
      new GraphWidget({
        title: 'Failed remediations by Control Id',
        left: [
          new MathExpression({
            expression: "SORT(SEARCH('{ASR,ControlId,Outcome} Outcome=\"FAILED\"', 'Sum'), SUM, ASC)",
            label: '',
            period: defaultDuration,
          }),
        ],
        view: GraphWidgetView.BAR,
        statistic: 'Sum',
      }),
      new GraphWidget({
        title: 'Remediation Failure Rate by Control Id',
        left: failureRateMetricsByControlId,
        leftAnnotations: [
          {
            value: remediationFailureAlarmThreshold.valueAsNumber,
            label: `S3.9 Failure Percentage >= threshold for 1 datapoints within 1 day`,
            color: Color.RED,
            visible: true,
          },
        ],
        view: GraphWidgetView.TIME_SERIES,
      }),
    );

    remediationDashboard.addWidgets(
      new TextWidget({
        markdown: `
## Runbook Assume Role Failures
This widget displays the frequency of the remediation lambda failing to assume the role necessary to remediate on a different account.

This may indicate that ASR is attempting to remediate on a spoke account that does not have ASR installed.
`,
        height: 3,
        width: 24,
      }),
      new GraphWidget({
        title: 'Runbook Assume Role Failures',
        left: [failedAssumeRoleMetric],
        leftAnnotations: [failedAssumeRoleAlarm.toAnnotation()],
        leftYAxis: {
          showUnits: false,
        },
      }),
    );

    remediationDashboard.addWidgets(
      new TextWidget({
        markdown: `
## Action Log
This widget displays AWS resource changes that ASR has conducted in member accounts.

The actions shown are based on CloudTrail management events in the member accounts. Actions are only reported if the member stack is deployed with "Create Action Log CloudTrail" set to "Yes".
`,
        height: 3,
        width: 24,
      }),
      new LogQueryWidget({
        logGroupNames: [props.actionLogLogGroupName],
        queryLines: [
          'fields @timestamp, eventSource, eventName, awsRegion, recipientAccountId, resources.0.ARN, @message',
          'sort @timestamp desc',
          'limit 20',
        ],
        height: 8,
        width: 24,
        title: 'CloudTrail Management Actions by ASR',
      }),
    );
  }

  private createAlarmsByControlId(
    scope: Construct,
    alarmThreshold: number,
    snsAlarmTopic: Topic,
    duration: Duration,
    controlIds: string[],
  ) {
    const metricsByControlId: IMetric[] = [];

    controlIds.forEach((controlId: string) => {
      const failuresByControlIdMetric = new Metric({
        namespace: 'ASR',
        metricName: 'RemediationOutcome',
        dimensionsMap: { Outcome: 'FAILED', ControlId: controlId },
        period: duration,
      });
      const successByControlIdMetric = new Metric({
        namespace: 'ASR',
        metricName: 'RemediationOutcome',
        dimensionsMap: { Outcome: 'SUCCESS', ControlId: controlId },
        period: duration,
      });

      const alphanumericControlId = controlId.replace(/\W/g, '');
      const failuresMetricName = `m1${alphanumericControlId}`;
      const successesMetricName = `m2${alphanumericControlId}`;
      const failurePercentage = new MathExpression({
        label: `${controlId} Failure Percentage`,
        period: duration,
        expression: `(${failuresMetricName} / (${failuresMetricName}+${successesMetricName})) * 100`,
        usingMetrics: {
          [failuresMetricName]: failuresByControlIdMetric,
          [successesMetricName]: successByControlIdMetric,
        },
      });
      metricsByControlId.push(failurePercentage);

      const remediationFailureAlarm = failurePercentage.createAlarm(scope, `${controlId}-remediation-failure`, {
        alarmName: `ASR-${controlId}-remediation-failure`,
        evaluationPeriods: 1,
        threshold: alarmThreshold,
        comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
        alarmDescription: `This alarm triggers when the percentage of remediation failures for ${controlId} reaches above the configured threshold. 
        This indicates that there may be a problem remediating this control ID in your AWS environment. Check the most recent failed execution of this control's runbook in the target account to identify potential issues.`,
        treatMissingData: TreatMissingData.NOT_BREACHING,
        datapointsToAlarm: 1,
      });

      setCondition(remediationFailureAlarm, this.enhancedAlarmsEnabled);
      remediationFailureAlarm.addAlarmAction(new SnsAction(snsAlarmTopic));

      addCfnGuardSuppression(remediationFailureAlarm, 'CFN_NO_EXPLICIT_RESOURCE_NAMES');
    });

    return metricsByControlId;
  }

  public getParameterIds(): string[] {
    return this.parameters.map((p) => p.logicalId);
  }

  public getParameterIdsAndLabels() {
    return this.parameters.reduce((a, p) => ({ ...a, [p.logicalId]: { default: p.logicalId } }), {});
  }

  public getCloudWatchMetricsParameterValue(): string {
    return this.useCloudWatchMetrics.valueAsString;
  }
}
