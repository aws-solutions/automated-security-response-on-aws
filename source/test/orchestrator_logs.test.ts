// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import * as cdk from 'aws-cdk-lib';
import { OrchLogStack } from '../lib/orchestrator-log-stack';
import { DefaultStackSynthesizer } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

function getTestStack(): cdk.Stack {
  const app = new cdk.App();
  const stack = new OrchLogStack(app, 'roles', {
    synthesizer: new DefaultStackSynthesizer({ generateBootstrapVersionRule: false }),
    description: 'test;',
    solutionId: 'SO0111',
    logGroupName: 'TestLogGroup',
  });
  return stack;
}
test('Global Roles Stack', () => {
  expect(Template.fromStack(getTestStack())).toMatchSnapshot();
});
