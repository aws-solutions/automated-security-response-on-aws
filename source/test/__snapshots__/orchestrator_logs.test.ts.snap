// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global Roles Stack 1`] = `
{
  "Conditions": {
    "EncryptedLogGroup": {
      "Fn::And": [
        {
          "Condition": "isNotGovCloud",
        },
        {
          "Fn::Equals": [
            {
              "Ref": "ReuseOrchestratorLogGroup",
            },
            "no",
          ],
        },
      ],
    },
    "UnencryptedLogGroup": {
      "Fn::And": [
        {
          "Fn::Not": [
            {
              "Condition": "isNotGovCloud",
            },
          ],
        },
        {
          "Fn::Equals": [
            {
              "Ref": "ReuseOrchestratorLogGroup",
            },
            "no",
          ],
        },
      ],
    },
    "isNotGovCloud": {
      "Fn::Not": [
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Partition",
            },
            "aws-us-gov",
          ],
        },
      ],
    },
  },
  "Description": "test;",
  "Parameters": {
    "KmsKeyArn": {
      "Description": "ARN of the KMS key to use to encrypt log data.",
      "Type": "String",
    },
    "ReuseOrchestratorLogGroup": {
      "AllowedValues": [
        "yes",
        "no",
      ],
      "Default": "no",
      "Description": "Reuse existing Orchestrator Log Group? Choose "yes" if you have deployed this stack before, otherwise "no"",
      "Type": "String",
    },
  },
  "Resources": {
    "OrchestratorLogsEFDFFA92": {
      "Condition": "UnencryptedLogGroup",
      "DeletionPolicy": "Retain",
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W84",
              "reason": "KmsKeyId is not supported in GovCloud.",
            },
          ],
        },
      },
      "Properties": {
        "LogGroupName": "TestLogGroup",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "OrchestratorLogsEncrypted072D6E38": {
      "Condition": "EncryptedLogGroup",
      "DeletionPolicy": "Retain",
      "Properties": {
        "KmsKeyId": {
          "Ref": "KmsKeyArn",
        },
        "LogGroupName": "TestLogGroup",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
